### ISSUE TITLE
Training bash for the video models

### ISSUE THREAD

**Issue by bryanyzhu** [pytorch/vision#1332]
Hi, the adoption of video in release 0.4.0 is very helpful. I was wondering will the traning bash (I mean the training recipe be made available as well)? Because right now, I can only find the `train.py` and the pre-trained models. If I want to train one myself, especially distributed training across multiple machines, the performance is not good. Maybe something is wrong with my warmup setup or learning rate setup. Is there any problem I should be aware of? Thank you for any help.

**Comment by fmassa** [comment]
Hi,

I've used a custom training script for launching the jobs on many machines, and the scripts are not yet open source.

For information, I trained those models on 8 machines, 64 GPUs. You might need to adapt the learning rates if you change the number of GPUs.

**Comment by bryanyzhu** [comment]
Thank you for your reply. So just adapt the learning rates, do I need to do something more? Because right now, I tried linear scaling of learning rates (e.g., 8 machines, then multiply lr by 8) and warm up. But the performance is worse than single machine results (the gap is about 3 percent accuracy).  Do you have any suggestions or comments? Thank you.

**Comment by fmassa** [comment]
Hey,

Sorry for the misleading information. The current implementation already multiplies by the number of GPUs in https://github.com/pytorch/vision/blob/367e8514c5d5e8528330081f25efeb17dd5e7ebd/references/video_classification/train.py#L211
so you don't need to perform learning-rate scaling yourself. Plus, warmup is already implemented as well.

Here are the arguments that I used for training (all parameters have been kept the same for different models):
```python
Namespace(apex=False, apex_opt_level='O1', batch_size=24,
    cache_dataset=True, clip_len=16, clips_per_video=5,
    data_path='/datasets01_101/kinetics/070618/',
    device='cuda', dist_backend='nccl',
    dist_url='file:///checkpoint/fmassa/submitit/f77095b1c3814a6f8e8478a5063746e4_init',
    distributed=True, epochs=45, gpu=0,
    lr=0.01, lr_gamma=0.1, lr_milestones=[20, 30, 40],
    lr_warmup_epochs=10, model='r2plus1d_18', momentum=0.9,
    output_dir='/checkpoint/fmassa/submitit/16259764',
    pretrained=False, print_freq=10, rank=0,
    resume='/checkpoint/fmassa/submitit/16233231/model_22.pth',
    start_epoch=0, sync_bn=False, test_only=False,
    weight_decay=0.0001, workers=10, world_size=64)
```
I run it with 8 machines, 64 GPUs, for 45 epochs. Best accuracy was obtained at epoch 40.

For info, I used the following versions:
```
torch version:  1.2.0.dev20190704
torchvision version:  0.3.0a0+f9788f3
```
but I believe there shouldn't be any BC-breaking change that would affect results.

How much worse are the results that you obtained in the single-machine setup?

Also ccing @bjuncek for thoughts.

**Comment by bryanyzhu** [comment]
Thanks a lot for the detailed answer, I will try the training gain using your recipe. 

My multi-machine result is about 2% worse than the single-machine result.

**Comment by fmassa** [comment]
> My multi-machine result is about 2% accuracy worse than the single-machine result.

And how much worse is it compared to the reference models that we provided?

**Comment by bjuncek** [comment]
@bryanyzhu Sorry for the delayed response.
Do you change any training parameters on multi-machine vs single machine training? 

---
also for reference, training procedure in the reference implementation closely follows [the reference implementation](https://arxiv.org/abs/1711.11248), but you would have to tune the hyperparameters a bit based on a) total batch size (across all machines) and b) total epoch size (clips_per_video).

**Comment by bryanyzhu** [comment]
Thank you very much for your comments. 

@fmassa My multi-machine result is 2.6% worse than the reference models. But I used 4 machines with 32 GPUs, maybe I need to tune some hyperparameters. 

@bjuncek Yes, I have changed the total batch size and total epoch size. I also changed learning rate before. I will check these params again.

**Comment by fmassa** [comment]
@bryanyzhu one question, what do you mean by
> My multi-machine result is about 2% accuracy worse than the single-machine result.

Do you mean that, for the same model, running the evaluation on 1 machine or 4 machines gives different results?

If that's the case, then I know what the issue might be.

We perform some padding on the dataset so that the number of clips is evenly-divisible by the number of GPUs, so that results should be slightly different.

But I wouldn't expect it to be so large, 2.6% is a lot.

**Comment by bryanyzhu** [comment]
@fmassa Thank you for the heads up. But what I mean is two models: one is trained in single-machine, the other is trained using multiple machines. The model trained using multiple machines gives worse results than model trained using a single machine.

For evaluation, I only use 1 machine, so should be ok.

**Comment by fmassa** [comment]
Ok, thanks for the info.

So from your previous messages it seems that the single-machine training gives 0.6% lower results than the reference models, is that right?

@bjuncek would this be considered within error or not?

**Comment by bjuncek** [comment]
This is possible - multiple machines allow for larger batch sizes and there is some loss of precision in gradient averaging. It could probably be solved with much finer HP tuning (like separate params for BN, ...) which we never really explored too much.  

I'd say that this is within the tolerance for all practical purposes.

**Comment by bryanyzhu** [comment]
@fmassa Yes, you are right. Single-machine training gives 0.6% lower results than the reference models.
